File: sdk/runanywhere-swift/Sources/RunAnywhere/Foundation/Logging/Models/SensitiveDataPolicy.swift
//
//  SensitiveDataPolicy.swift
//  RunAnywhere SDK
//
//  Defines policies for handling sensitive data in logs
//

import Foundation

/// Policy for handling sensitive data in logs
public enum SensitiveDataPolicy {
    /// Data is not sensitive - can be logged locally and remotely
    case none

    /// Data contains sensitive information - log locally only in debug mode, never remotely
    case sensitive

    /// Data is critical - never log the actual content, only log category/type
    case critical

    /// Completely redact - don't log at all
    case redacted
}

/// Categories of sensitive data for classification
public enum SensitiveDataCategory {
    // Authentication & Credentials
    case apiKey
    case bearerToken
    case password
    case credential

    // User Content
    case userPrompt
    case generatedResponse
    case conversationHistory
    case systemPrompt
    case userMessage

    // Personal Information
    case userId
    case email
    case phoneNumber
    case deviceIdentifier

    // Configuration & System
    case modelPath
    case fileSystemPath
    case internalURL
    case errorDetails

    // Business Logic
    case costCalculation
    case routingDecision
    case performanceMetric

    var defaultPolicy: SensitiveDataPolicy {
        switch self {
        case .apiKey, .bearerToken, .password, .credential:
            return .redacted
        case .userPrompt, .generatedResponse, .conversationHistory, .systemPrompt, .userMessage:
            return .critical
        case .userId, .email, .phoneNumber, .deviceIdentifier:
            return .critical
        case .modelPath, .fileSystemPath, .internalURL:
            return .sensitive
        case .errorDetails, .costCalculation, .routingDecision, .performanceMetric:
            return .sensitive
        }
    }

    var sanitizedPlaceholder: String {
        switch self {
        case .apiKey: return "[API_KEY]"
        case .bearerToken: return "[TOKEN]"
        case .password: return "[PASSWORD]"
        case .credential: return "[CREDENTIAL]"
        case .userPrompt: return "[USER_PROMPT]"
        case .generatedResponse: return "[GENERATED_RESPONSE]"
        case .conversationHistory: return "[CONVERSATION]"
        case .systemPrompt: return "[SYSTEM_PROMPT]"
        case .userMessage: return "[USER_MESSAGE]"
        case .userId: return "[USER_ID]"
        case .email: return "[EMAIL]"
        case .phoneNumber: return "[PHONE]"
        case .deviceIdentifier: return "[DEVICE_ID]"
        case .modelPath: return "[MODEL_PATH]"
        case .fileSystemPath: return "[FILE_PATH]"
        case .internalURL: return "[URL]"
        case .errorDetails: return "[ERROR_DETAILS]"
        case .costCalculation: return "[COST_DATA]"
        case .routingDecision: return "[ROUTING_DATA]"
        case .performanceMetric: return "[PERFORMANCE_DATA]"
        }
    }
}

/// Metadata key for marking sensitive data
public struct LogMetadataKeys {
    public static let sensitiveDataPolicy = "__sensitive_data_policy"
    public static let sensitiveDataCategory = "__sensitive_data_category"
    public static let isUserContent = "__is_user_content"
    public static let containsPII = "__contains_pii"
    public static let sanitized = "__sanitized"
}

File: sdk/runanywhere-flutter/lib/foundation/logging/models/sensitive_data_policy.dart
/// Policy for handling sensitive data in logs
///
/// Aligned with iOS: Sources/RunAnywhere/Foundation/Logging/Models/SensitiveDataPolicy.swift
enum SensitiveDataPolicy {
  /// Data is not sensitive - can be logged locally and remotely
  none,

  /// Data contains sensitive information - log locally only in debug mode, never remotely
  sensitive,

  /// Data is critical - never log the actual content, only log category/type
  critical,

  /// Completely redact - don't log at all
  redacted,
}

/// Categories of sensitive data for classification
enum SensitiveDataCategory {
  // Authentication & Credentials
  apiKey,
  bearerToken,
  password,
  credential,

  // User Content
  userPrompt,
  generatedResponse,
  conversationHistory,
  systemPrompt,
  userMessage,

  // Personal Information
  userId,
  email,
  phoneNumber,
  deviceIdentifier,

  // Configuration & System
  modelPath,
  fileSystemPath,
  internalURL,
  errorDetails,

  // Business Logic
  costCalculation,
  routingDecision,
  performanceMetric,
}

/// Extension for SensitiveDataCategory
extension SensitiveDataCategoryExtension on SensitiveDataCategory {
  /// Get the default policy for this category
  SensitiveDataPolicy get defaultPolicy {
    switch (this) {
      case SensitiveDataCategory.apiKey:
      case SensitiveDataCategory.bearerToken:
      case SensitiveDataCategory.password:
      case SensitiveDataCategory.credential:
        return SensitiveDataPolicy.redacted;

      case SensitiveDataCategory.userPrompt:
      case SensitiveDataCategory.generatedResponse:
      case SensitiveDataCategory.conversationHistory:
      case SensitiveDataCategory.systemPrompt:
      case SensitiveDataCategory.userMessage:
        return SensitiveDataPolicy.critical;

      case SensitiveDataCategory.userId:
      case SensitiveDataCategory.email:
      case SensitiveDataCategory.phoneNumber:
      case SensitiveDataCategory.deviceIdentifier:
        return SensitiveDataPolicy.critical;

      case SensitiveDataCategory.modelPath:
      case SensitiveDataCategory.fileSystemPath:
      case SensitiveDataCategory.internalURL:
        return SensitiveDataPolicy.sensitive;

      case SensitiveDataCategory.errorDetails:
      case SensitiveDataCategory.costCalculation:
      case SensitiveDataCategory.routingDecision:
      case SensitiveDataCategory.performanceMetric:
        return SensitiveDataPolicy.sensitive;
    }
  }

  /// Get the sanitized placeholder for this category
  String get sanitizedPlaceholder {
    switch (this) {
      case SensitiveDataCategory.apiKey:
        return '[API_KEY]';
      case SensitiveDataCategory.bearerToken:
        return '[TOKEN]';
      case SensitiveDataCategory.password:
        return '[PASSWORD]';
      case SensitiveDataCategory.credential:
        return '[CREDENTIAL]';
      case SensitiveDataCategory.userPrompt:
        return '[USER_PROMPT]';
      case SensitiveDataCategory.generatedResponse:
        return '[GENERATED_RESPONSE]';
      case SensitiveDataCategory.conversationHistory:
        return '[CONVERSATION]';
      case SensitiveDataCategory.systemPrompt:
        return '[SYSTEM_PROMPT]';
      case SensitiveDataCategory.userMessage:
        return '[USER_MESSAGE]';
      case SensitiveDataCategory.userId:
        return '[USER_ID]';
      case SensitiveDataCategory.email:
        return '[EMAIL]';
      case SensitiveDataCategory.phoneNumber:
        return '[PHONE]';
      case SensitiveDataCategory.deviceIdentifier:
        return '[DEVICE_ID]';
      case SensitiveDataCategory.modelPath:
        return '[MODEL_PATH]';
      case SensitiveDataCategory.fileSystemPath:
        return '[FILE_PATH]';
      case SensitiveDataCategory.internalURL:
        return '[URL]';
      case SensitiveDataCategory.errorDetails:
        return '[ERROR_DETAILS]';
      case SensitiveDataCategory.costCalculation:
        return '[COST_DATA]';
      case SensitiveDataCategory.routingDecision:
        return '[ROUTING_DATA]';
      case SensitiveDataCategory.performanceMetric:
        return '[PERFORMANCE_DATA]';
    }
  }
}

/// Metadata keys for marking sensitive data
class LogMetadataKeys {
  LogMetadataKeys._();

  static const String sensitiveDataPolicy = '__sensitive_data_policy';
  static const String sensitiveDataCategory = '__sensitive_data_category';
  static const String isUserContent = '__is_user_content';
  static const String containsPII = '__contains_pii';
  static const String sanitized = '__sanitized';
}
